MATLAB/mimo_driver.py

```diff
@@ -0,0 +1,669 @@
+        [ue.set_corr() for ue in self.ue_obj]
+
+        rx_data = np.empty((num_frames, n_users, numRxSyms, n_samps), dtype=np.complex64)
+        good_frame_id = 0
+
+        for frame in range(num_frames):
+            all_triggered = False
+            good_signal = False
+            amp = 0
+            #if frame > 0:
+            old_triggers = []
+            triggers = []
+            for u in range(self.n_users):
+                old_triggers.append(0)
+            for i in range(max_try):
+                self.bs_trigger()
+                # Receive Data
+                rx_data_frame_tmp = [ue.recv_stream_tdd() for ue in self.ue_obj]
+                rx_data_frame = np.array(rx_data_frame_tmp)
+                rx_data_frame = rx_data_frame[:,0:2]
+
+                #if i == 0:
+                #    savemat("./testtest.mat", {'rxdataOBCH': rx_data_frame})
+
+                # Verify Data
+                if self.n_ue_chan > 1:
+                    both_channels = True
+                else:
+                    both_channels = False
+                good_signal = self.verify_signal_quality(rx_data_frame_tmp, self.n_users, both_channels)
+
+                amp = np.mean(np.abs(rx_data_frame[0][0]))
+                #good_signal = amp > 0.001
+                triggers = [ue.sdr_gettriggers() for ue in self.ue_obj]
+                new_triggers = []
+                zip_triggers = zip(triggers, old_triggers)
+                for triggers_i, old_triggers_i in zip_triggers:
+                    new_triggers.append(triggers_i - old_triggers_i)
+                print("triggers = {}, Amplitude = {}".format(new_triggers, amp))
+                all_triggered = True
+                for u in range(n_users):
+                    if (new_triggers[u] == 0):
+                        all_triggered = False
+                if all_triggered and good_signal:
+                    break
+
+            print("frame = {}, tries = {}, all_triggred = {}, good_signal = {}, amp = {}".format(frame, i + 1, all_triggered, good_signal, amp))
+            if all_triggered and good_signal:
+                if n_users == 1 and n_bs_antenna == 1:
+                    rx_data[good_frame_id, :, :, :] = np.reshape(np.array(rx_data_frame[0][0]), (n_bs_antenna, numRxSyms, n_samps))
+                else:
+                    rx_data[good_frame_id, :, :, :] = np.reshape(np.array(rx_data_frame[0][0]), (n_users, numRxSyms, n_samps))  # FIXME: Will require changes for multiple users (similar to uplink but for UEs)
+                good_frame_id = good_frame_id + 1
+                #tmptmp = np.array(rx_data_frame[0])
+                #plt.plot(np.abs(tmptmp))
+
+        #plt.show()
+        [ue.unset_corr() for ue in self.ue_obj]
+        self.reset_frame()
+
+        return rx_data, good_frame_id, numRxSyms
+
+
+    def txrx_dl_sound(self, tx_data_mat_re, tx_data_mat_im, num_frames, nsamps_pad=160, max_try=30, python_mode=False):
```

Comment:
This function is very similar to the `txrx_downlink`? Can they be merged?

Reply:
they are different but we could do that. I'll leave that as a TODO. don't want to break anything after all the testing. I'll add it to my next change